File: 60-LeetCode-Problems/2week/Word-Ladder.py
from collections import deque
from typing import Deque, List, Set


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        wordSet: Set[str] = set(wordList)
        temp: Deque = deque()
        temp.append((beginWord, 1))
        s = set()
        s.add(beginWord)

        while temp:
            now, cnt = temp.popleft()

            if now == endWord:
                return cnt

            for c1 in range(len(now)):
                for c2 in range(97, 123):
                    temp_word = now[:c1] + chr(c2) + now[c1 + 1:]
                    if temp_word in wordSet and temp_word not in s:
                        s.add(temp_word)
                        temp.append((temp_word, cnt + 1))
        return 0

File: 60-LeetCode-Problems/2week/Word-Ladder.java
class Solution {
    
    private Map<String, List<String>> initCases (List<String> wordList, int maxWordLength) 
    {
        Map<String, List<String>> cases = new HashMap<>();
       
        
        for (int i = 0; i < wordList.size(); i++) {
            String cur = wordList.get(i);
            for( int j =0;j<maxWordLength;j++)
        {
           String key= cur.substring(0,j)+"_"+cur.substring(j+1,maxWordLength);
            if(cases.containsKey(key))
                {
                cases.get(key).add(cur);
                }
            else
                {
                cases.put(key, new ArrayList<String>());
                 cases.get(key).add(cur);
                }
            }
        }
        System.out.println(cases);
        return cases;
    }
    
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if (!wordList.contains(endWord)) {
             return 0;
            }
        
        int cnt = 0;
        int maxWordLength = wordList.get(0).length();
        Map<String, List<String>> availableCases = initCases(wordList, maxWordLength);
        Queue<String> wordQueue = new LinkedList<>(); 
        HashSet<String> visited = new HashSet<>();
        wordQueue.add(beginWord);
         
        while( wordQueue.size()>0)
    {   
        int queueSize =wordQueue.size();
        cnt++;
        
        while (queueSize --> 0)
        {
          String currWord= wordQueue.poll();
          for(int i =0; i<currWord.length();i++)
          {
            List<String> nextChars= availableCases.get(currWord.substring(0,i)+"_"+currWord.substring(i+1,currWord.length()));
             System.out.println(nextChars);
            if(nextChars!=null && !nextChars.isEmpty() )
            {
                for(String s: nextChars)
                {
                    if(s.equals(endWord))
                        return cnt+1;
                    
                   if( !visited.contains(s))
                    {    
                         visited.add(s);
                         wordQueue.add(s);
                    }
                }
            }
          }
        }

    }
    return 0;
    }
}